File: frontend/src/lib/fillerRemoval.js
/**
 * Feature #14 — filler-word & silence removal (frontend detection + the
 * caption-remap mirror).
 *
 * Detection produces candidate cut spans ([start,end] in clip-local seconds);
 * the store seeds cutSpans from these when the feature is enabled, the user
 * restores individual spans, and the effective list rides the wire to the
 * backend (which renders the cuts and remaps captions with the SAME remap
 * math mirrored in services/filler_removal.py — keep them in lockstep).
 *
 * WYSIWYG note: the editor preview plays the UNCUT media, so cuts show as
 * struck-through transcript words; the effect lands on export (Ganesh's gate).
 */

// Conservative filler set — romanized + common Telugu hesitation sounds. Kept
// small on purpose: the user reviews every struck-through word and restores
// false positives, so over-detection is worse than under-detection.
export const FILLER_WORDS = new Set([
  "um", "uh", "umm", "uhh", "uhm", "hmm", "hm", "erm", "err", "ah", "aa",
  "aah", "eh", "mm", "mmm", "like", "so",
  "ఆ", "ఆఁ", "అ", "ఉమ్", "హ్మ్", "అమ్మ",
]);

const norm = (s) =>
  String(s || "").trim().toLowerCase().replace(/[.,!?;:"'“”‘’()]/g, "");

// Silence = a gap between consecutive words longer than minGap. We keep a
// small pad on each side so speech onsets/tails aren't clipped; the removed
// span is the interior of the gap.
export const DEFAULT_MIN_GAP = 0.6;
export const SILENCE_PAD = 0.1;

// words: [{start,end,text,...}] clip-local. Returns candidate spans with meta
// so the UI can label them: {start, end, kind: 'filler'|'silence', wordIndex?}.
export function detectRemovableSpans(
  words,
  duration,
  { minGap = DEFAULT_MIN_GAP, pad = SILENCE_PAD } = {}
) {
  const spans = [];
  const ws = words || [];
  for (let i = 0; i < ws.length; i++) {
    const w = ws[i];
    if (typeof w.start !== "number" || typeof w.end !== "number") continue;
    // Filler word → remove its own span.
    if (FILLER_WORDS.has(norm(w.text))) {
      spans.push({ start: w.start, end: w.end, kind: "filler", wordIndex: i });
    }
    // Silence gap before the NEXT word.
    const next = ws[i + 1];
    if (next && typeof next.start === "number") {
      const gap = next.start - w.end;
      if (gap > minGap) {
        const s = w.end + pad;
        const e = next.start - pad;
        if (e > s) spans.push({ start: round3(s), end: round3(e), kind: "silence" });
      }
    }
  }
  return mergeSpans(spans);
}

// Merge overlapping/adjacent spans (sorted). Meta collapses to the first
// contributor's kind; the render only needs the [start,end] union.
export function mergeSpans(spans) {
  const sorted = [...(spans || [])]
    .filter((s) => s && s.end > s.start)
    .sort((a, b) => a.start - b.start);
  const out = [];
  for (const s of sorted) {
    const last = out[out.length - 1];
    if (last && s.start <= last.end + 1e-6) {
      last.end = Math.max(last.end, s.end);
    } else {
      out.push({ ...s });
    }
  }
  return out;
}

// Total removed duration strictly BEFORE time t.
export function removedBefore(t, spans) {
  let removed = 0;
  for (const s of spans || []) {
    if (s.end <= t) removed += s.end - s.start;
    else if (s.start < t) removed += t - s.start; // t lands inside this span
  }
  return removed;
}

// Is a word fully or majority inside any cut span (→ dropped from captions)?
export function wordInSpans(word, spans) {
  const mid = (word.start + word.end) / 2;
  return (spans || []).some((s) => mid >= s.start && mid < s.end);
}

// Remap a clip-local time onto the post-cut output timeline.
export function remapTimeAfterCuts(t, spans) {
  return round3(Math.max(0, t - removedBefore(t, spans)));
}

// Drop words inside cuts; shift survivors onto the post-cut timeline. Mirror
// of services/filler_removal.py::apply_cuts_to_words — WYSIWYG contract.
export function applyCutsToWords(words, spans) {
  if (!spans || !spans.length) return words || [];
  const out = [];
  for (const w of words || []) {
    if (wordInSpans(w, spans)) continue;
    out.push({
      ...w,
      start: remapTimeAfterCuts(w.start, spans),
      end: remapTimeAfterCuts(w.end, spans),
    });
  }
  return out;
}

// [start,end] pairs for the wire (backend render). Strips meta.
export function spansToPairs(spans) {
  return (spans || []).map((s) => [round3(s.start), round3(s.end)]);
}

const round3 = (n) => Math.round(n * 1000) / 1000;

File: services/filler_removal.py
# -*- coding: utf-8 -*-
"""
Feature #14 — filler-word & silence removal (render + caption remap).

The frontend detects candidate cut spans (fillers + silence gaps), lets the
user restore individual ones, and sends the effective cut list as
[[start, end], ...] clip-local seconds. This module:
  - renders the cuts (worker._apply_cuts uses build_select_filters) via a
    single FFmpeg select/aselect + setpts/asetpts pass, and
  - remaps caption word timings onto the post-cut timeline
    (apply_cuts_to_words) — the MIRROR of frontend lib/fillerRemoval.js.

Depends on feature #1: cut spans are in the clip's refined (cut-file) time
base, the same base captions already use.
"""


def _norm_spans(spans):
    """Sanitize → sorted, merged [(start, end)] with end > start."""
    clean = []
    for s in spans or []:
        try:
            a, b = float(s[0]), float(s[1])
        except (TypeError, ValueError, IndexError):
            continue
        if b > a:
            clean.append((a, b))
    clean.sort()
    merged = []
    for a, b in clean:
        if merged and a <= merged[-1][1] + 1e-6:
            merged[-1] = (merged[-1][0], max(merged[-1][1], b))
        else:
            merged.append((a, b))
    return merged


def keep_spans(cut_spans, duration):
    """Complement of the cut spans over [0, duration] — the parts we KEEP."""
    cuts = _norm_spans(cut_spans)
    keeps = []
    cursor = 0.0
    for a, b in cuts:
        a = max(0.0, min(a, duration))
        b = max(0.0, min(b, duration))
        if a > cursor:
            keeps.append((round(cursor, 3), round(a, 3)))
        cursor = max(cursor, b)
    if cursor < duration:
        keeps.append((round(cursor, 3), round(duration, 3)))
    return [(a, b) for a, b in keeps if b > a]


def build_cut_filtergraph(cut_spans, duration):
    """FFmpeg -filter_complex string that keeps the non-cut parts and concats
    them, or None when nothing is cut.

    Uses per-segment trim/atrim + concat (NOT select/aselect): select drops
    video frames fine but aselect+asetpts left the audio full-length in
    testing, desyncing the result. trim/atrim+concat cuts BOTH streams
    identically and re-bases each segment's PTS, so audio and video stay
    locked. Output labels are [outv]/[outa] for the worker's -map.
    """
    keeps = keep_spans(cut_spans, duration)
    # No cuts, or a single keep spanning the whole clip → caller passes through.
    if not keeps or (len(keeps) == 1 and keeps[0][0] <= 0 and keeps[0][1] >= duration):
        return None
    parts, labels = [], []
    for i, (a, b) in enumerate(keeps):
        parts.append(f"[0:v]trim={a}:{b},setpts=PTS-STARTPTS[v{i}]")
        parts.append(f"[0:a]atrim={a}:{b},asetpts=PTS-STARTPTS[a{i}]")
        labels.append(f"[v{i}][a{i}]")
    concat = "".join(labels) + f"concat=n={len(keeps)}:v=1:a=1[outv][outa]"
    return ";".join(parts) + ";" + concat


def total_removed(cut_spans, duration):
    cuts = _norm_spans(cut_spans)
    return round(sum(min(b, duration) - max(a, 0.0)
                     for a, b in cuts if min(b, duration) > max(a, 0.0)), 3)


# ── caption remap (mirror of frontend lib/fillerRemoval.js) ──────────────────

def _removed_before(t, cuts):
    removed = 0.0
    for a, b in cuts:
        if b <= t:
            removed += b - a
        elif a < t:
            removed += t - a
    return removed


def remap_time_after_cuts(t, cut_spans):
    cuts = _norm_spans(cut_spans)
    return round(max(0.0, t - _removed_before(t, cuts)), 3)


def _word_in_cuts(word, cuts):
    mid = (word["start"] + word["end"]) / 2.0
    return any(a <= mid < b for a, b in cuts)


def apply_cuts_to_words(words, cut_spans):
    """Drop words inside cuts; shift survivors onto the post-cut timeline.
    Mirror of frontend applyCutsToWords — preview and burn must agree."""
    cuts = _norm_spans(cut_spans)
    if not cuts:
        return words
    out = []
    for w in words:
        if _word_in_cuts(w, cuts):
            continue
        nw = dict(w)
        nw["start"] = remap_time_after_cuts(w["start"], cut_spans)
        nw["end"] = remap_time_after_cuts(w["end"], cut_spans)
        out.append(nw)
    return out
